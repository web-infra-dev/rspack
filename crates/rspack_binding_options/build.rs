use std::env;

fn main() {
  // Rebuild binding options if and only if it's built for crate `node_binding`
  if env::var("OUT_DIR")
    .expect("should exist")
    .contains("node_binding")
  {
    println!("cargo:rerun-if-changed=../node_binding");
  }
}
