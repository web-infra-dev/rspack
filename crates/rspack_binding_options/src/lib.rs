#![feature(try_blocks)]
#![feature(let_chains)]
mod options;
mod plugins;
pub use options::*;
