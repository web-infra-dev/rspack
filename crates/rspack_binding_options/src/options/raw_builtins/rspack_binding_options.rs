use std::sync::Arc;

use derivative::Derivative;
use napi::{bindgen_prelude::Either3, Either, Env, JsFunction};
use napi_derive::napi;
use rspack_napi_shared::{
  get_napi_env,
  threadsafe_function::{ThreadsafeFunction, ThreadsafeFunctionCallMode},
  JsRegExp, NapiResultExt,
};
use rspack_plugin_devtool::{
  Append, DevtoolPluginOptions, FallbackModuleFilenameTemplate, ModuleFilenameTemplate,
};
use serde::Deserialize;

type RawAppend = Either3<String, bool, JsFunction>;

#[inline]
fn default_append() -> Append {
  Append::Default
}

fn normalize_raw_append(raw: RawAppend) -> Append {
  match raw {
    Either3::A(str) => Append::String(str),
    Either3::B(_) => Append::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      Append::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

type RawRule = Either<String, JsRegExp>;
type RawRules = Either<RawRule, Vec<RawRule>>;

type RawFallbackModuleFilenameTemplate = Either3<String, bool, JsFunction>;

#[inline]
fn default_fallback_module_filename_template() -> FallbackModuleFilenameTemplate {
  FallbackModuleFilenameTemplate::Disabled
}

fn normalize_raw_fallback_module_filename_template(
  raw: RawFallbackModuleFilenameTemplate,
) -> FallbackModuleFilenameTemplate {
  match raw {
    Either3::A(str) => FallbackModuleFilenameTemplate::String(str),
    Either3::B(_) => FallbackModuleFilenameTemplate::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      FallbackModuleFilenameTemplate::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

type RawModuleFilenameTemplate = Either3<String, bool, JsFunction>;

#[inline]
fn default_module_filename_template() -> ModuleFilenameTemplate {
  ModuleFilenameTemplate::Disabled
}

fn normalize_raw_module_filename_template(
  raw: RawModuleFilenameTemplate,
) -> ModuleFilenameTemplate {
  match raw {
    Either3::A(str) => ModuleFilenameTemplate::String(str),
    Either3::B(_) => ModuleFilenameTemplate::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      ModuleFilenameTemplate::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

#[derive(Derivative, Deserialize)]
#[serde(rename_all = "camelCase")]
#[napi(object)]
#[derivative(Debug)]
pub struct RawSourceMapDevToolPluginOptions {
  #[serde(skip_deserializing)]
  #[napi(ts_type = "(false | null) | string | Function")]
  #[derivative(Debug = "ignore")]
  pub append: Option<RawAppend>,
  pub columns: Option<bool>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub exclude: Option<RawRules>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | false | Function")]
  #[derivative(Debug = "ignore")]
  pub fallback_module_filename_template: Option<RawFallbackModuleFilenameTemplate>,
  pub file_context: Option<String>,
  #[napi(ts_type = "(false | null) | string")]
  pub filename: Option<String>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub include: Option<RawRules>,
  pub module: Option<bool>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | false | Function")]
  #[derivative(Debug = "ignore")]
  pub module_filename_template: Option<RawModuleFilenameTemplate>,
  pub namespace: Option<String>,
  pub no_sources: Option<bool>,
  pub public_path: Option<String>,
  pub source_root: Option<String>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub test: Option<RawRules>,
}

impl From<RawSourceMapDevToolPluginOptions> for DevtoolPluginOptions {
  fn from(raw_opts: RawSourceMapDevToolPluginOptions) -> Self {
    let append = raw_opts
      .append
      .map_or(default_append(), |name| normalize_raw_append(name));

    DevtoolPluginOptions {
      append,
      inline: todo!(),
      namespace: todo!(),
      columns: todo!(),
      no_sources: todo!(),
      public_path: raw_opts.public_path,
    }
  }
}
