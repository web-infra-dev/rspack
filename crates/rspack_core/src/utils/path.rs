use std::path::{Component, Path};
use sugar_path::{self, PathSugar};
pub fn normalize_path(path: &str, root: &str) -> String {
  let res = Path::new(&root)
    .relative(Path::new(&path))
    .to_string_lossy()
    .to_string();

  res
}

pub fn uri_to_chunk_name(root: &str, uri: &str) -> String {
  println!("root {:?}", root);
  let path = Path::new(uri);
  let mut relatived = Path::new(root).relative(path);
  let ext = relatived
    .extension()
    .and_then(|ext| ext.to_str())
    .unwrap_or("")
    .to_string();
  relatived.set_extension("");
  let mut name = relatived
    .components()
    .filter(|com| matches!(com, Component::Normal(_)))
    .filter_map(|seg| seg.as_os_str().to_str())
    .intersperse("_")
    .fold(String::new(), |mut acc, seg| {
      acc.push_str(seg);
      acc
    });
  name.push('_');
  name.push_str(&ext);
  name
}
