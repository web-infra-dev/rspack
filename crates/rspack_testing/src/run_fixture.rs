use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

use cargo_rst::{helper::make_relative_from, rst::RstBuilder};
use insta::{assert_snapshot, Settings};
use itertools::Itertools;
use rspack_binding_options::{RawOptions, RawOptionsApply};
use rspack_core::{BoxPlugin, Compiler, CompilerOptions};
use rspack_fs::AsyncNativeFileSystem;
use rspack_tracing::enable_tracing_by_env;

use crate::{eval_raw::evaluate_to_json, test_config::TestConfig};

pub type MutTestOptionsFn = dyn FnMut(&mut Vec<BoxPlugin>, &mut CompilerOptions);

pub fn apply_from_fixture(fixture_path: &Path) -> (CompilerOptions, Vec<BoxPlugin>) {
  let js_config = fixture_path.join("test.config.js");
  if js_config.exists() {
    let raw = evaluate_to_json(&js_config);
    let raw: RawOptions = serde_json::from_slice(&raw).expect("ok");
    let mut plugins = Vec::new();
    let compiler_options = raw.apply(&mut plugins).expect("should be ok");
    return (compiler_options, plugins);
  }
  let json_config = fixture_path.join("test.config.json");
  let test_config = TestConfig::from_config_path(&json_config);
  test_config.apply(fixture_path.to_path_buf())
}
#[tokio::main]
pub async fn test_fixture_html(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".html"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_js(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".js") && !s.contains("runtime.js"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_css(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".css"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_css_modules(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".css") || (s.ends_with(".js") && !s.contains("runtime.js")),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_insta(
  fixture_path: &Path,
  stats_filter: &dyn Fn(&str) -> bool,
  mut_settings: Box<MutTestOptionsFn>,
) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(fixture_path, stats_filter, mut_settings, None).await
}

#[tokio::main]
pub async fn test_fixture(
  fixture_path: &Path,
  mut_settings: Box<MutTestOptionsFn>,
  snapshot_name: Option<String>,
) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".js") && !s.contains("runtime.js"),
    mut_settings,
    snapshot_name,
  )
  .await
}

/// You can mutate the `Settings` of insta and  `CompilerOptions` of rspack via `mut_settings` mut closure
pub async fn test_fixture_share(
  fixture_path: &Path,
  stats_filter: &dyn Fn(&str) -> bool,
  mut mut_settings: Box<MutTestOptionsFn>,
  snapshot_name: Option<String>,
) -> Compiler<AsyncNativeFileSystem> {
  enable_tracing_by_env(&std::env::var("TRACE").ok().unwrap_or_default(), "stdout");

  let snapshot_name = snapshot_name.unwrap_or("output".to_string());

  let mut settings = Settings::clone_current();
  settings.set_snapshot_path(Path::new(fixture_path).join("snapshot"));
  settings.set_omit_expression(true);
  settings.set_prepend_module_to_snapshot(false);

  let (mut options, mut plugins) = apply_from_fixture(fixture_path);

  mut_settings(&mut plugins, &mut options);

  // clean output
  if options.output.path.exists() {
    std::fs::remove_dir_all(&options.output.path).expect("should remove output");
  }
  let mut compiler = Compiler::new(options, plugins, AsyncNativeFileSystem);

  compiler
    .build()
    .await
    .unwrap_or_else(|e| panic!("failed to compile in fixture {fixture_path:?}, {e:#?}"));
  let assets = compiler.compilation.assets();

  let content = assets
    .iter()
    .filter_map(|(filename, asset)| {
      if stats_filter(filename) {
        println!("{:?}", settings.description());
        let content = asset
          .get_source()
          .map(|x| x.source().to_string())
          .unwrap_or(String::from("this is an empty asset"));
        let tag = Path::new(filename)
          .extension()
          .map(|x| x.to_string_lossy())
          .unwrap_or(std::borrow::Cow::Borrowed("txt"));
        Some(format!("```{tag} title={filename}\n{content}\n```"))
      } else {
        None
      }
    })
    .sorted()
    .join("\n\n");
  settings.bind(|| {
    assert_snapshot!(snapshot_name.as_str(), content);
  });
  let stats = compiler.compilation.get_stats();
  let warnings = stats.get_warnings();
  let errors = stats.get_errors();
  if !warnings.is_empty() && errors.is_empty() {
    println!(
      "Warning to compile in fixture {:?}, warnings: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    )
  }
  if !errors.is_empty() {
    panic!(
      "Failed to compile in fixture {:?}, errors: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    );
  }
  compiler
}

fn read_dir_reverse(path: &PathBuf) -> Vec<String> {
  let mut result = vec![];
  if let Ok(changed_dir) = std::fs::read_dir(path) {
    changed_dir.for_each(|entry| {
      let entry = entry.expect("TODO:");
      if entry.path().is_file() {
        result.push(entry.path().as_os_str().to_string_lossy().to_string());
      }
      if entry.path().is_dir() {
        result.extend(read_dir_reverse(&entry.path()))
      }
    })
  }
  result
}

enum FsOptionEnum {
  Change(Vec<u8>),
  Create,
  Remove(Vec<u8>),
}

#[tokio::main]
pub async fn test_rebuild_fixture(
  fixture_path: &Path,
  cb: Option<Box<dyn FnOnce(Compiler<AsyncNativeFileSystem>)>>,
) {
  enable_tracing_by_env(&std::env::var("TRACE").ok().unwrap_or_default(), "stdout");

  let (options, plugins) = apply_from_fixture(fixture_path);
  // clean output
  if options.output.path.exists() {
    std::fs::remove_dir_all(&options.output.path).expect("should remove output");
  }
  let mut compiler = Compiler::new(options, plugins, AsyncNativeFileSystem);
  compiler
    .build()
    .await
    .unwrap_or_else(|e| panic!("failed to compile in fixture {fixture_path:?}, {e:#?}"));

  let mut files_map: HashMap<String, FsOptionEnum> = HashMap::new();
  let changed_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("changed")).into_iter());
  let removed_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("removed")).into_iter());
  let created_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("created")).into_iter());

  let mut created_files: HashSet<String> = created_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("created/", "", 1);
      files_map.insert(target_path.clone(), FsOptionEnum::Create);
      let created_context = std::fs::read(file_path.clone());
      std::fs::write(
        target_path.clone(),
        created_context.expect("changed file do not read"),
      )
      .expect("TODO: panic message");
      target_path
    })
    .collect();
  let changed_files: HashSet<String> = changed_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("changed/", "", 1);
      let target_context = std::fs::read(target_path.clone());
      files_map.insert(
        target_path.clone(),
        FsOptionEnum::Change(target_context.expect("change file not found")),
      );
      let new_content = std::fs::read(file_path);
      std::fs::write(
        target_path.clone(),
        new_content.expect("changed file do not read"),
      )
      .expect("TODO: panic message");
      target_path
    })
    .collect();
  let removed_files: HashSet<String> = removed_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("removed/", "", 1);
      let target_context = std::fs::read(target_path.clone());
      files_map.insert(
        target_path.clone(),
        FsOptionEnum::Remove(target_context.expect("change file not found")),
      );
      target_path
    })
    .collect();
  created_files.extend(changed_files.into_iter());
  let changed_files = created_files;
  compiler
    .rebuild(changed_files, removed_files)
    .await
    .unwrap_or_else(|e| panic!("failed to rebuild in fixture {fixture_path:?}, {e:#?}"));

  files_map
    .iter()
    .for_each(|(old_path, old_content)| match old_content {
      FsOptionEnum::Create => {
        std::fs::remove_file(old_path).expect("TODO: panic message");
      }
      FsOptionEnum::Change(old_content) | FsOptionEnum::Remove(old_content) => {
        std::fs::write(old_path.clone(), old_content).expect("TODO: panic message");
      }
    });
  let stats = compiler.compilation.get_stats();
  let output_name = make_relative_from(&compiler.options.output.path, fixture_path);
  let rst = RstBuilder::default()
    .fixture(PathBuf::from(fixture_path))
    .actual(output_name)
    .build()
    .expect("TODO:");

  let errors = stats.get_errors();
  if !errors.is_empty() {
    panic!(
      "Failed to compile in fixture {:?}, errors: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    );
  }
  rst.assert();

  if let Some(cb) = cb {
    cb(compiler);
  }
}
