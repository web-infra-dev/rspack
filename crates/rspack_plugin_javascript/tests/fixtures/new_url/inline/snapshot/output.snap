---
source: crates/rspack_testing/src/run_fixture.rs
---
```js title=main.js
(self['webpackChunkwebpack'] = self['webpackChunkwebpack'] || []).push([["main"], {
"./index.js": function (__unused_webpack_module, exports, __webpack_require__) {
const imgSrc = new URL(/* asset import */__webpack_require__(/* ./react.svg */"./react.svg"), __webpack_require__.b);
const imgSrc2 = __webpack_require__(/* ./vue.svg */"./vue.svg");
const img = new Image();
img.src = imgSrc.href;
img.src = imgSrc2;
},
"./react.svg": function (module, exports, __webpack_require__) {
module.exports = "data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA4NDEuOSA1OTUuMyI+PGcgZmlsbD0iIzYxREFGQiI+PHBhdGggZD0iTTY2Ni4zIDI5Ni41YzAtMzIuNS00MC43LTYzLjMtMTAzLjEtODIuNCAxNC40LTYzLjYgOC0xMTQuMi0yMC4yLTEzMC40LTYuNS0zLjgtMTQuMS01LjYtMjIuNC01LjZ2MjIuM2M0LjYgMCA4LjMuOSAxMS40IDIuNiAxMy42IDcuOCAxOS41IDM3LjUgMTQuOSA3NS43LTEuMSA5LjQtMi45IDE5LjMtNS4xIDI5LjQtMTkuNi00LjgtNDEtOC41LTYzLjUtMTAuOS0xMy41LTE4LjUtMjcuNS0zNS4zLTQxLjYtNTAgMzIuNi0zMC4zIDYzLjItNDYuOSA4NC00Ni45Vjc4Yy0yNy41IDAtNjMuNSAxOS42LTk5LjkgNTMuNi0zNi40LTMzLjgtNzIuNC01My4yLTk5LjktNTMuMnYyMi4zYzIwLjcgMCA1MS40IDE2LjUgODQgNDYuNi0xNCAxNC43LTI4IDMxLjQtNDEuMyA0OS45LTIyLjYgMi40LTQ0IDYuMS02My42IDExLTIuMy0xMC00LTE5LjctNS4yLTI5LTQuNy0zOC4yIDEuMS02Ny45IDE0LjYtNzUuOCAzLTEuOCA2LjktMi42IDExLjUtMi42Vjc4LjVjLTguNCAwLTE2IDEuOC0yMi42IDUuNi0yOC4xIDE2LjItMzQuNCA2Ni43LTE5LjkgMTMwLjEtNjIuMiAxOS4yLTEwMi43IDQ5LjktMTAyLjcgODIuMyAwIDMyLjUgNDAuNyA2My4zIDEwMy4xIDgyLjQtMTQuNCA2My42LTggMTE0LjIgMjAuMiAxMzAuNCA2LjUgMy44IDE0LjEgNS42IDIyLjUgNS42IDI3LjUgMCA2My41LTE5LjYgOTkuOS01My42IDM2LjQgMzMuOCA3Mi40IDUzLjIgOTkuOSA1My4yIDguNCAwIDE2LTEuOCAyMi42LTUuNiAyOC4xLTE2LjIgMzQuNC02Ni43IDE5LjktMTMwLjEgNjItMTkuMSAxMDIuNS00OS45IDEwMi41LTgyLjN6bS0xMzAuMi02Ni43Yy0zLjcgMTIuOS04LjMgMjYuMi0xMy41IDM5LjUtNC4xLTgtOC40LTE2LTEzLjEtMjQtNC42LTgtOS41LTE1LjgtMTQuNC0yMy40IDE0LjIgMi4xIDI3LjkgNC43IDQxIDcuOXptLTQ1LjggMTA2LjVjLTcuOCAxMy41LTE1LjggMjYuMy0yNC4xIDM4LjItMTQuOSAxLjMtMzAgMi00NS4yIDItMTUuMSAwLTMwLjItLjctNDUtMS45LTguMy0xMS45LTE2LjQtMjQuNi0yNC4yLTM4LTcuNi0xMy4xLTE0LjUtMjYuNC0yMC44LTM5LjggNi4yLTEzLjQgMTMuMi0yNi44IDIwLjctMzkuOSA3LjgtMTMuNSAxNS44LTI2LjMgMjQuMS0zOC4yIDE0LjktMS4zIDMwLTIgNDUuMi0yIDE1LjEgMCAzMC4yLjcgNDUgMS45IDguMyAxMS45IDE2LjQgMjQuNiAyNC4yIDM4IDcuNiAxMy4xIDE0LjUgMjYuNCAyMC44IDM5LjgtNi4zIDEzLjQtMTMuMiAyNi44LTIwLjcgMzkuOXptMzIuMy0xM2M1LjQgMTMuNCAxMCAyNi44IDEzLjggMzkuOC0xMy4xIDMuMi0yNi45IDUuOS00MS4yIDggNC45LTcuNyA5LjgtMTUuNiAxNC40LTIzLjcgNC42LTggOC45LTE2LjEgMTMtMjQuMXpNNDIxLjIgNDMwYy05LjMtOS42LTE4LjYtMjAuMy0yNy44LTMyIDkgLjQgMTguMi43IDI3LjUuNyA5LjQgMCAxOC43LS4yIDI3LjgtLjctOSAxMS43LTE4LjMgMjIuNC0yNy41IDMyem0tNzQuNC01OC45Yy0xNC4yLTIuMS0yNy45LTQuNy00MS03LjkgMy43LTEyLjkgOC4zLTI2LjIgMTMuNS0zOS41IDQuMSA4IDguNCAxNiAxMy4xIDI0IDQuNyA4IDkuNSAxNS44IDE0LjQgMjMuNHpNNDIwLjcgMTYzYzkuMyA5LjYgMTguNiAyMC4zIDI3LjggMzItOS0uNC0xOC4yLS43LTI3LjUtLjctOS40IDAtMTguNy4yLTI3LjguNyA5LTExLjcgMTguMy0yMi40IDI3LjUtMzJ6bS03NCA1OC45Yy00LjkgNy43LTkuOCAxNS42LTE0LjQgMjMuNy00LjYgOC04LjkgMTYtMTMgMjQtNS40LTEzLjQtMTAtMjYuOC0xMy44LTM5LjggMTMuMS0zLjEgMjYuOS01LjggNDEuMi03Ljl6bS05MC41IDEyNS4yYy0zNS40LTE1LjEtNTguMy0zNC45LTU4LjMtNTAuNiAwLTE1LjcgMjIuOS0zNS42IDU4LjMtNTAuNiA4LjYtMy43IDE4LTcgMjcuNy0xMC4xIDUuNyAxOS42IDEzLjIgNDAgMjIuNSA2MC45LTkuMiAyMC44LTE2LjYgNDEuMS0yMi4yIDYwLjYtOS45LTMuMS0xOS4zLTYuNS0yOC0xMC4yek0zMTAgNDkwYy0xMy42LTcuOC0xOS41LTM3LjUtMTQuOS03NS43IDEuMS05LjQgMi45LTE5LjMgNS4xLTI5LjQgMTkuNiA0LjggNDEgOC41IDYzLjUgMTAuOSAxMy41IDE4LjUgMjcuNSAzNS4zIDQxLjYgNTAtMzIuNiAzMC4zLTYzLjIgNDYuOS04NCA0Ni45LTQuNS0uMS04LjMtMS0xMS4zLTIuN3ptMjM3LjItNzYuMmM0LjcgMzguMi0xLjEgNjcuOS0xNC42IDc1LjgtMyAxLjgtNi45IDIuNi0xMS41IDIuNi0yMC43IDAtNTEuNC0xNi41LTg0LTQ2LjYgMTQtMTQuNyAyOC0zMS40IDQxLjMtNDkuOSAyMi42LTIuNCA0NC02LjEgNjMuNi0xMSAyLjMgMTAuMSA0LjEgMTkuOCA1LjIgMjkuMXptMzguNS02Ni43Yy04LjYgMy43LTE4IDctMjcuNyAxMC4xLTUuNy0xOS42LTEzLjItNDAtMjIuNS02MC45IDkuMi0yMC44IDE2LjYtNDEuMSAyMi4yLTYwLjYgOS45IDMuMSAxOS4zIDYuNSAyOC4xIDEwLjIgMzUuNCAxNS4xIDU4LjMgMzQuOSA1OC4zIDUwLjYtLjEgMTUuNy0yMyAzNS42LTU4LjQgNTAuNnpNMzIwLjggNzguNHoiLz48Y2lyY2xlIGN4PSI0MjAuOSIgY3k9IjI5Ni41IiByPSI0NS43Ii8+PHBhdGggZD0iTTUyMC41IDc4LjF6Ii8+PC9nPjwvc3ZnPg==";},
"./vue.svg": function (module, exports, __webpack_require__) {
module.exports = "data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiBhcmlhLWhpZGRlbj0idHJ1ZSIgcm9sZT0iaW1nIiBjbGFzcz0iaWNvbmlmeSBpY29uaWZ5LS1sb2dvcyIgd2lkdGg9IjM3LjA3IiBoZWlnaHQ9IjM2IiBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSJ4TWlkWU1pZCBtZWV0IiB2aWV3Qm94PSIwIDAgMjU2IDE5OCI+PHBhdGggZmlsbD0iIzQxQjg4MyIgZD0iTTIwNC44IDBIMjU2TDEyOCAyMjAuOEwwIDBoOTcuOTJMMTI4IDUxLjJMMTU3LjQ0IDBoNDcuMzZaIj48L3BhdGg+PHBhdGggZmlsbD0iIzQxQjg4MyIgZD0ibTAgMGwxMjggMjIwLjhMMjU2IDBoLTUxLjJMMTI4IDEzMi40OEw1MC41NiAwSDBaIj48L3BhdGg+PHBhdGggZmlsbD0iIzM1NDk1RSIgZD0iTTUwLjU2IDBMMTI4IDEzMy4xMkwyMDQuOCAwaC00Ny4zNkwxMjggNTEuMkw5Ny45MiAwSDUwLjU2WiI+PC9wYXRoPjwvc3ZnPg==";},

},function(__webpack_require__) {
var __webpack_exec__ = function(moduleId) { return __webpack_require__(__webpack_require__.s = moduleId) }
var __webpack_exports__ = (__webpack_exec__("./index.js"));

}
]);
```
```js title=runtime.js
(function() {
var __webpack_modules__ = {

}
// The module cache
 var __webpack_module_cache__ = {};
function __webpack_require__(moduleId) {
// Check if module is in cache
        var cachedModule = __webpack_module_cache__[moduleId];
        if (cachedModule !== undefined) {
      return cachedModule.exports;
      }
      // Create a new module (and put it into the cache)
      var module = (__webpack_module_cache__[moduleId] = {
       exports: {}
      });
      // Execute the module function
      __webpack_modules__[moduleId](module, module.exports, __webpack_require__);
// Return the exports of the module
 return module.exports;

}
// expose the modules object (__webpack_modules__)
 __webpack_require__.m = __webpack_modules__;
// webpack/runtime/has_own_property
(function() {
__webpack_require__.o = function (obj, prop) {
	return Object.prototype.hasOwnProperty.call(obj, prop);
};

})();
// webpack/runtime/on_chunk_loaded
(function() {
var deferred = [];
__webpack_require__.O = function (result, chunkIds, fn, priority) {
	if (chunkIds) {
		priority = priority || 0;
		for (var i = deferred.length; i > 0 && deferred[i - 1][2] > priority; i--)
			deferred[i] = deferred[i - 1];
		deferred[i] = [chunkIds, fn, priority];
		return;
	}
	var notFulfilled = Infinity;
	for (var i = 0; i < deferred.length; i++) {
		var chunkIds = deferred[i][0],
			fn = deferred[i][1],
			priority = deferred[i][2];
		var fulfilled = true;
		for (var j = 0; j < chunkIds.length; j++) {
			if (
				(priority & (1 === 0) || notFulfilled >= priority) &&
				Object.keys(__webpack_require__.O).every(function (key) {
					return __webpack_require__.O[key](chunkIds[j]);
				})
			) {
				chunkIds.splice(j--, 1);
			} else {
				fulfilled = false;
				if (priority < notFulfilled) notFulfilled = priority;
			}
		}
		if (fulfilled) {
			deferred.splice(i--, 1);
			var r = fn();
			if (r !== undefined) result = r;
		}
	}
	return result;
};

})();
// webpack/runtime/jsonp_chunk_loading
(function() {
__webpack_require__.b = document.baseURI || self.location.href;
var installedChunks = {"runtime": 0,};
__webpack_require__.O.j = function (chunkId) {
	return installedChunks[chunkId] === 0;
};
// install a JSONP callback for chunk loading
var webpackJsonpCallback = function (parentChunkLoadingFunction, data) {
	var chunkIds = data[0],
	moreModules = data[1],
	runtime = data[2];
	// add "moreModules" to the modules object,
	// then flag all "chunkIds" as loaded and fire callback
	var moduleId,
		chunkId,
		i = 0;
	if (chunkIds.some(function(id) { return installedChunks[id] !== 0 })) {
		for (moduleId in moreModules) {
			if (__webpack_require__.o(moreModules, moduleId)) {
				__webpack_require__.m[moduleId] = moreModules[moduleId];
			}
		}
		if (runtime) var result = runtime(__webpack_require__);
	}
	if (parentChunkLoadingFunction) parentChunkLoadingFunction(data);
	for (; i < chunkIds.length; i++) {
		chunkId = chunkIds[i];
		if (
			__webpack_require__.o(installedChunks, chunkId) &&
			installedChunks[chunkId]
		) {
			installedChunks[chunkId][0]();
		}
		installedChunks[chunkId] = 0;
	}
	return __webpack_require__.O(result);
};

var chunkLoadingGlobal = self['webpackChunkwebpack'] = self['webpackChunkwebpack'] || [];
chunkLoadingGlobal.forEach(webpackJsonpCallback.bind(null, 0));
chunkLoadingGlobal.push = webpackJsonpCallback.bind(
	null,
	chunkLoadingGlobal.push.bind(chunkLoadingGlobal)
);

})();

})()

```
